src/analysis/processing/qgsalgorithmrastercalculator.cpp

```diff
@@ -0,0 +1,192 @@
+{
+  const QList< QgsMapLayer * > layers = parameterAsLayerList( parameters, QStringLiteral( "INPUT" ), context );
+
+  for ( const QgsMapLayer *layer : std::as_const( layers ) )
+  {
+    QgsMapLayer *clonedLayer { layer->clone() };
+    mLayers << clonedLayer;
+  }
+
+  if ( mLayers.isEmpty() )
+  {
+    feedback->reportError( QObject::tr( "No layers selected" ), false );
+    return false;
+  }
+
+  return true;
+}
+
+
+QVariantMap QgsRasterCalculatorAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
+{
+  QgsCoordinateReferenceSystem crs;
+  if ( parameters.value( QStringLiteral( "CRS" ) ).isValid() )
+  {
+    crs = parameterAsCrs( parameters, QStringLiteral( "CRS" ), context );
+  }
+  else
+  {
+    crs = mLayers.at( 0 )->crs();
+  }
+
+  QgsRectangle bbox;
+  if ( parameters.value( QStringLiteral( "EXTENT" ) ).isValid() )
+  {
+    bbox = parameterAsExtent( parameters, QStringLiteral( "EXTENT" ), context, crs );
+  }
+  else
+  {
+    bbox = QgsProcessingUtils::combineLayerExtents( mLayers, crs, context );
+  }
+
+  double minCellSize = 1e9;
+
+  QVector< QgsRasterCalculatorEntry > entries;
+  for ( QgsMapLayer *layer : mLayers )
+  {
+    QgsRasterLayer *rLayer = static_cast<QgsRasterLayer *>( layer );
+    if ( !rLayer )
+    {
+      continue;
+    }
+
+    const int nBands = rLayer->dataProvider()->bandCount();
+    for ( int i = 0; i < nBands; ++i )
+    {
+      QgsRasterCalculatorEntry entry;
+      entry.ref = QStringLiteral( "%1@%2" ).arg( rLayer->name() ).arg( i + 1 );
+      entry.raster = rLayer;
+      entry.bandNumber = i + 1;
+      entries << entry;
+    }
+
+    QgsRectangle ext = rLayer->extent();
+    if ( rLayer->crs().authid() != crs.authid() )
+    {
+      QgsCoordinateTransform ct( rLayer->crs(), crs, context.transformContext() );
+      ext = ct.transformBoundingBox( ext );
+    }
+
+    double cellSize = ( ext.xMaximum() - ext.xMinimum() ) / rLayer->width();
+    if ( cellSize < minCellSize )
+    {
+      minCellSize = cellSize;
+    }
+  }
+
+  double cellSize = parameterAsDouble( parameters, QStringLiteral( "CELL_SIZE" ), context );
+  if ( cellSize == 0 )
+  {
+    cellSize = minCellSize;
+  }
+
+  const QString expression = parameterAsExpression( parameters, QStringLiteral( "EXPRESSION" ), context );
+  const QString outputFile = parameterAsOutputLayer( parameters, QStringLiteral( "OUTPUT" ), context );
+  const QFileInfo fi( outputFile );
+  const QString outputFormat = QgsRasterFileWriter::driverForExtension( fi.suffix() );
+
+  double width = std::round( ( bbox.xMaximum() - bbox.xMinimum() ) / cellSize );
+  double height = std::round( ( bbox.yMaximum() - bbox.yMinimum() ) / cellSize );
+
+  QgsRasterCalculator calc( expression, outputFile, outputFormat, bbox, crs, width, height, entries, context.transformContext() );
+  QgsRasterCalculator::Result result = calc.processCalculation( feedback );
+  switch ( result )
```

Comment:
There's a leak here -- we'll need to delete all the layers before returning from this function:

```suggestion
  QgsRasterCalculator::Result result = calc.processCalculation( feedback );
  qDeleteAll( mLayers );
  mLayers.clear();
  switch ( result )
```

Reply:
Fixed, thanks.